packages/react/src/tanstackrouter.ts

```diff
@@ -0,0 +1,167 @@
+            ...routeMatchToParamSpanAttributes(lastMatch),
+          },
+        });
+      }
+
+      if (instrumentNavigation) {
+        // The onBeforeNavigate hook is called at the very beginning of a navigation and is only called once per navigation, even when the user is redirected
+        castRouterInstance.subscribe('onBeforeNavigate', onBeforeNavigateArgs => {
+          // onBeforeNavigate is called during pageloads. We can avoid creating navigation spans by comparing the states of the to and from arguments.
+          if (onBeforeNavigateArgs.toLocation.state === onBeforeNavigateArgs.fromLocation.state) {
+            return;
+          }
+
+          const onResolvedMatchedRoutes = castRouterInstance.matchRoutes(
+            onBeforeNavigateArgs.toLocation.pathname,
+            onBeforeNavigateArgs.toLocation.search,
+            { preload: false, throwOnError: false },
+          );
+
+          const onBeforeNavigateLastMatch = onResolvedMatchedRoutes[onResolvedMatchedRoutes.length - 1];
+
+          const navigationLocation = WINDOW.location;
+          const navigationSpan = startBrowserTracingNavigationSpan(client, {
+            name: onBeforeNavigateLastMatch ? onBeforeNavigateLastMatch.routeId : navigationLocation.pathname,
+            attributes: {
+              [SEMANTIC_ATTRIBUTE_SENTRY_OP]: 'navigation',
+              [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: 'auto.navigation.react.tanstack_router',
+              [SEMANTIC_ATTRIBUTE_SENTRY_SOURCE]: onBeforeNavigateLastMatch ? 'route' : 'url',
+            },
+          });
+
+          // In case the user is redirected during navigation we want to update the span with the right value.
+          const unsubscribeOnResolved = castRouterInstance.subscribe('onResolved', onResolvedArgs => {
+            unsubscribeOnResolved();
+            if (navigationSpan) {
+              const onResolvedMatchedRoutes = castRouterInstance.matchRoutes(
+                onResolvedArgs.toLocation.pathname,
+                onResolvedArgs.toLocation.search,
+                { preload: false, throwOnError: false },
+              );
+
+              const onResolvedLastMatch = onResolvedMatchedRoutes[onResolvedMatchedRoutes.length - 1];
+
+              if (onResolvedLastMatch) {
+                navigationSpan.updateName(onResolvedLastMatch.routeId);
+                navigationSpan.setAttribute(SEMANTIC_ATTRIBUTE_SENTRY_SOURCE, 'route');
+                navigationSpan.setAttributes(routeMatchToParamSpanAttributes(onResolvedLastMatch));
+              }
+            }
+          });
+        });
+      }
+    },
+  };
+}
+
+function routeMatchToParamSpanAttributes(match: VendoredTanstackRouterRouteMatch | undefined): Record<string, string> {
+  if (!match) {
+    return {};
+  }
+
+  const paramAttributes: Record<string, string> = {};
+  for (const key of Object.keys(match.params)) {
+    paramAttributes[`params.${key}`] = match.params[key];
```

Comment:
should we only set this with `sendDefaultPii`?

Also, I think we should call this `url.path.params.X` because the OTEL attribute for path is `url.path`.

Reply:
> should we only set this with sendDefaultPii?

I would opt against it since all of the params are inside the URL (which we send by default) anyhow.

If for some reason, we would be sending additional data after a patch, I would guard it behind sendDefaultPii.

 > Also, I think we should call this url.path.params.X because the OTEL attribute for path is url.path
 
 👍